File: mobile/android/app/src/main/kotlin/com/aetherride/aetherride_mobile/BoschLdiProto.kt
package com.aetherride.aetherride_mobile

/**
 * Bosch LDI LiveData protobuf (Spec V1.0, Apache-2.0). Sparse fields.
 * Does not invent SoC / power — omitted keys stay absent.
 */
internal data class BoschLdiSnapshot(
    var speedKmh: Double? = null,
    var cadenceRpm: Double? = null,
    var riderPowerW: Double? = null,
    var ambientBrightness: Double? = null,
    var batterySocPercent: Double? = null,
    var odometerKm: Double? = null,
    var lightStatus: Boolean? = null,
    var systemLock: Boolean? = null,
    var chargerConnected: Boolean? = null,
    var bikeNotDriving: Boolean? = null,
) {
    fun merge(frame: BoschLdiSnapshot) {
        frame.speedKmh?.let { speedKmh = it }
        frame.cadenceRpm?.let { cadenceRpm = it }
        frame.riderPowerW?.let { riderPowerW = it }
        frame.ambientBrightness?.let { ambientBrightness = it }
        frame.batterySocPercent?.let { batterySocPercent = it }
        frame.odometerKm?.let { odometerKm = it }
        frame.lightStatus?.let { lightStatus = it }
        frame.systemLock?.let { systemLock = it }
        frame.chargerConnected?.let { chargerConnected = it }
        frame.bikeNotDriving?.let { bikeNotDriving = it }
    }

    fun toEventMap(): Map<String, Any?> {
        val m = HashMap<String, Any?>()
        speedKmh?.let { m["speedKmh"] = it }
        cadenceRpm?.let { m["cadenceRpm"] = it }
        odometerKm?.let { m["odometerKm"] = it }
        lightStatus?.let { m["lightStatus"] = it }
        ambientBrightness?.let { m["ambientBrightness"] = it }
        systemLock?.let { m["systemLock"] = it }
        bikeNotDriving?.let { m["bikeNotDriving"] = it }
        chargerConnected?.let { m["chargerConnected"] = it }
        batterySocPercent?.let { m["batterySocPercent"] = it }
        riderPowerW?.let { m["riderPowerW"] = it }
        m["timestampMs"] = System.currentTimeMillis()
        return m
    }
}

internal object BoschLdiProto {
    fun decode(bytes: ByteArray): BoschLdiSnapshot {
        val out = BoschLdiSnapshot()
        var i = 0
        fun u8(): Int {
            if (i >= bytes.size) throw IllegalArgumentException("truncated")
            return bytes[i++].toInt() and 0xff
        }
        fun varint(): Long {
            var shift = 0
            var n = 0L
            while (true) {
                val b = u8()
                n = n or ((b and 0x7f).toLong() shl shift)
                if (b and 0x80 == 0) return n
                shift += 7
                if (shift > 63) throw IllegalArgumentException("varint")
            }
        }
        fun skip(wire: Int) {
            when (wire) {
                0 -> varint()
                1 -> i += 8
                2 -> {
                    val n = varint().toInt()
                    i += n
                }
                5 -> i += 4
                else -> throw IllegalArgumentException("wire $wire")
            }
        }
        while (i < bytes.size) {
            val tag = varint().toInt()
            val field = tag ushr 3
            val wire = tag and 7
            if (wire != 0) {
                skip(wire)
                continue
            }
            val n = varint()
            when (field) {
                1 -> out.speedKmh = n / 100.0
                2 -> {
                    val u = n and 0xffffffffL
                    out.cadenceRpm = if (u <= 0x7fffffffL) {
                        u.toDouble()
                    } else {
                        (u - 0x100000000L).toDouble()
                    }
                }
                5 -> out.riderPowerW = n.toDouble()
                9 -> out.ambientBrightness = n / 1000.0
                10 -> out.batterySocPercent = n.coerceIn(0L, 100L).toDouble()
                12 -> out.odometerKm = n / 1000.0
                17 -> out.lightStatus = n.toInt() == 2
                21 -> out.systemLock = n != 0L
                22 -> out.chargerConnected = n != 0L
                25 -> out.bikeNotDriving = n != 0L
            }
        }
        return out
    }
}

File: mobile/lib/domain/ble/bosch_ldi_proto.dart
/// Decoder for Bosch LDI `LiveData` protobuf (Spec V1.0, Apache-2.0).
/// Sparse notifications merge in the native plugin; this parses one frame.
class BoschLdiFrame {
  const BoschLdiFrame({
    this.speedKmh,
    this.cadenceRpm,
    this.riderPowerW,
    this.ambientBrightnessLux,
    this.batterySocPercent,
    this.timeUnixS,
    this.odometerKm,
    this.lightOn,
    this.systemLock,
    this.chargerConnected,
    this.lightReserve,
    this.diagnosisActive,
    this.bikeNotDriving,
  });

  final double? speedKmh;
  final double? cadenceRpm;
  final double? riderPowerW;
  final double? ambientBrightnessLux;
  final double? batterySocPercent;
  final int? timeUnixS;
  final double? odometerKm;
  final bool? lightOn;
  final bool? systemLock;
  final bool? chargerConnected;
  final bool? lightReserve;
  final bool? diagnosisActive;
  final bool? bikeNotDriving;

  bool get hasLiveMetrics =>
      speedKmh != null ||
      cadenceRpm != null ||
      riderPowerW != null ||
      batterySocPercent != null;
}

class _Reader {
  _Reader(this.bytes);
  final List<int> bytes;
  int i = 0;

  bool get done => i >= bytes.length;

  int u8() {
    if (i >= bytes.length) throw const FormatException('truncated');
    return bytes[i++];
  }

  int varint() {
    var shift = 0;
    var n = 0;
    while (true) {
      final b = u8();
      n |= (b & 0x7f) << shift;
      if ((b & 0x80) == 0) return n;
      shift += 7;
      if (shift > 63) throw const FormatException('varint overflow');
    }
  }

  List<int> skip(int wire) {
    switch (wire) {
      case 0:
        varint();
        return const [];
      case 1:
        i += 8;
        return const [];
      case 2:
        final n = varint();
        i += n;
        return const [];
      case 5:
        i += 4;
        return const [];
      default:
        throw FormatException('wire $wire');
    }
  }
}

int _sint32FromVarint(int n) {
  final u = n & 0xffffffff;
  if (u <= 0x7fffffff) return u;
  return u - 0x100000000;
}

/// Parse one `com.bosch.ebike.LiveData` payload. Unknown fields ignored.
BoschLdiFrame decodeBoschLdiFrame(List<int> bytes) {
  final r = _Reader(bytes);
  double? speedKmh;
  double? cadenceRpm;
  double? riderPowerW;
  double? ambientBrightnessLux;
  double? batterySocPercent;
  int? timeUnixS;
  double? odometerKm;
  bool? lightOn;
  bool? systemLock;
  bool? chargerConnected;
  bool? lightReserve;
  bool? diagnosisActive;
  bool? bikeNotDriving;

  while (!r.done) {
    final tag = r.varint();
    final field = tag >> 3;
    final wire = tag & 7;
    if (wire != 0) {
      r.skip(wire);
      continue;
    }
    final n = r.varint();
    switch (field) {
      case 1:
        speedKmh = n / 100.0;
      case 2:
        cadenceRpm = _sint32FromVarint(n).toDouble();
      case 5:
        riderPowerW = n.toDouble();
      case 9:
        ambientBrightnessLux = n / 1000.0;
      case 10:
        batterySocPercent = n.clamp(0, 100).toDouble();
      case 11:
        timeUnixS = n;
      case 12:
        odometerKm = n / 1000.0;
      case 17:
        lightOn = n == 2;
      case 21:
        systemLock = n != 0;
      case 22:
        chargerConnected = n != 0;
      case 23:
        lightReserve = n != 0;
      case 24:
        diagnosisActive = n != 0;
      case 25:
        bikeNotDriving = n != 0;
    }
  }

  return BoschLdiFrame(
    speedKmh: speedKmh,
    cadenceRpm: cadenceRpm,
    riderPowerW: riderPowerW,
    ambientBrightnessLux: ambientBrightnessLux,
    batterySocPercent: batterySocPercent,
    timeUnixS: timeUnixS,
    odometerKm: odometerKm,
    lightOn: lightOn,
    systemLock: systemLock,
    chargerConnected: chargerConnected,
    lightReserve: lightReserve,
    diagnosisActive: diagnosisActive,
    bikeNotDriving: bikeNotDriving,
  );
}

/// Encode a few fields for tests (proto3 varint).
List<int> encodeBoschLdiTestFrame({
  int? speedHundredths,
  int? cadenceRpm,
  int? riderPowerW,
  int? batterySoc,
  int? odometerM,
}) {
  final out = <int>[];
  void field(int id, int value) {
    var tag = (id << 3);
    while (tag > 0x7f) {
      out.add((tag & 0x7f) | 0x80);
      tag >>= 7;
    }
    out.add(tag);
    var n = value;
    if (n < 0) n = n & 0xffffffff;
    while (n > 0x7f) {
      out.add((n & 0x7f) | 0x80);
      n >>= 7;
    }
    out.add(n);
  }

  if (speedHundredths != null) field(1, speedHundredths);
  if (cadenceRpm != null) field(2, cadenceRpm);
  if (riderPowerW != null) field(5, riderPowerW);
  if (batterySoc != null) field(10, batterySoc);
  if (odometerM != null) field(12, odometerM);
  return out;
}
